### DISCUSSION TITLE
Another dumb question :)

### DISCUSSION THREAD

**Discussion by inhahe** [emscripten-core/emscripten#22978] [General]
Is this valid? `  Module.ccall('mainloop', null, [], []).then(() =>` (insert code here)...Chrome complains that 'then' can't read properties of undefined. If I can't do that, how do I call something only after mainloop finishes?

(I know the normal thing to do is to use EMSCRIPTEN_SET_MAINLOOP or whatever it's called, but I'm not updating the canvas from C++/the wasm program, I want to do it in JavaScript (reading info generated by the wasm), so I'm not sure if I should try to do it that way..)

**Comment by inhahe** [comment]
While I'm here, I might as well ask about the soundness of my overall strategy:
```
      function mainLoop()
      {
        Module.ccall('mainloop', null, [], []).then(() =>
        {
          var pixels = HEAPU8.subarray(pixelsptr, pixelsptr + w * h * 4);
          //pixels = new Uint8Array(Module.HEAPU8.buffer, pixelsptr, pixelsptr + w * h * 4);
//          imagedata = [...pixels];
          //imagedata = pixels.concat(); //which is more efficient?
          //ctx.putImageData(imagadata, 0, 0);
          ctx.putImageData(pixels, 0, 0);
          console.log("here");
        })
        requestAnimationFrame(mainLoop);
      }
    </script>
```
1. Will calling mainLoop from within mainLoop (in the `requestAnimationFrame(mainLoop);` line) repeatedly result in a stack overflow, or will it do tail recursion or something? If so, how else should I do this? Or I guess since it's a callback it won't actually be called from within mainLoop?
2. Can I directly putImageData the subarray like in the (not commented out) code? Or do I need to convert it to some other type of array? Or do I need to use something created by ctx.createImageData?
3. I don't know what the () after "then(" does, do I have to put something between the parentheses?

**Comment by sbc100** [comment]
If `mainloop` is just a normal C function you don't need any kind of async here you can just perform any actions you want directly after the call.  e.g.:

```
var mainLoopRrtn = Module.ccall('mainloop', null, [], []);
doSomethingAfter();
```

Doing `requestAnimationFrame(mainLoop);` is generally fine even withing `mainLoop` itself.  That line never actually call `mainLoop` it just puts it in the queue to be called on the next frame.
